base/set.jl

```diff
@@ -123,16 +123,45 @@ julia> unique([1; 2; 2; 6])
  6
 ```
 """
-function unique(C)
-    out = Vector{eltype(C)}()
-    seen = Set{eltype(C)}()
-    for x in C
+function unique(itr)
+    T = _default_eltype(typeof(itr))
+    out = Vector{T}()
+    seen = Set{T}()
+    i = start(itr)
+    if done(itr, i)
+        return out
+    end
+    x, i = next(itr, i)
+    if !isleaftype(T)
+        S = typeof(x)
+        return _unique_from(itr, S[x], push!(Set{S}(), x), i)
```

Comment:
Not really. The `Set` constructor is defined for iterables, so that won't work if `x` is not iterable. But `Set{S}((x,))` should work.